From 0142df58e2c12d40df7d46cdec11ae76cd135fed Mon Sep 17 00:00:00 2001
From: Jun Nie <jun.nie@linaro.org>
Date: Fri, 27 Jul 2018 09:15:15 +0800
Subject: [PATCH 2/2] rpi3: Enable boot kernel from fit image

Enable booting kernel from fit image with select configs
and specify boot script image node in fit image, bootscr
node in mbl case.

Code that reusing dtb in firmware is disabled.

Signed-off-by: Jun Nie <jun.nie@linaro.org>
---
 board/raspberrypi/rpi/rpi.c | 6 ++++++
 include/configs/rpi.h       | 8 ++++++--
 2 files changed, 12 insertions(+), 2 deletions(-)

diff --git a/board/raspberrypi/rpi/rpi.c b/board/raspberrypi/rpi/rpi.c
index 9e0abdd..fec90c9 100644
--- a/board/raspberrypi/rpi/rpi.c
+++ b/board/raspberrypi/rpi/rpi.c
@@ -322,6 +322,7 @@ int dram_init_banksize(void)
 #endif
 #endif
 
+#ifndef CONFIG_FIT_SIGNATURE
 static void set_fdtfile(void)
 {
 	const char *fdtfile;
@@ -347,6 +348,7 @@ static void set_fdt_addr(void)
 
 	env_set_hex("fdt_addr", fw_dtb_pointer);
 }
+#endif
 
 /*
  * Prevent relocation from stomping on a firmware provided FDT blob.
@@ -429,8 +431,10 @@ static void set_serial_number(void)
 
 int misc_init_r(void)
 {
+#ifndef CONFIG_FIT_SIGNATURE
 	set_fdt_addr();
 	set_fdtfile();
+#endif
 	set_usbethaddr();
 #ifdef CONFIG_ENV_VARS_UBOOT_RUNTIME_CONFIG
 	set_board_info();
@@ -506,6 +510,7 @@ int board_init(void)
 	return bcm2835_power_on_module(BCM2835_MBOX_POWER_DEVID_USB_HCD);
 }
 
+#ifndef CONFIG_FIT_SIGNATURE
 /*
  * If the firmware passed a device tree use it for U-Boot.
  */
@@ -515,6 +520,7 @@ void *board_fdt_blob_setup(void)
 		return NULL;
 	return (void *)fw_dtb_pointer;
 }
+#endif
 
 int ft_board_setup(void *blob, bd_t *bd)
 {
diff --git a/include/configs/rpi.h b/include/configs/rpi.h
index 77d2d54..a0b1cb3 100644
--- a/include/configs/rpi.h
+++ b/include/configs/rpi.h
@@ -14,6 +14,7 @@
 #endif
 
 /* Architecture, CPU, etc.*/
+#define CONFIG_GZIP_COMPRESSED
 
 /* Use SoC timer for AArch32, but architected timer for AArch64 */
 #ifndef CONFIG_ARM64
@@ -186,7 +187,10 @@
 	"dhcpuboot=usb start; dhcp u-boot.uimg; bootm\0" \
 	ENV_DEVICE_SETTINGS \
 	ENV_MEM_LAYOUT_SETTINGS \
-	BOOTENV
-
+	BOOTENV \
+	"boot_a_script="                                                  \
+		"load ${devtype} ${devnum}:${distro_bootpart} "           \
+			"${scriptaddr} ${prefix}${script}; "              \
+		"source ${scriptaddr}:bootscr\0"
 
 #endif
